packages/clay-core/src/tree-view/useTree.ts

```diff
@@ -0,0 +1,191 @@
+}
+
+export function useTree<T>(props: ITreeProps<T>): ITreeState<T> {
+	const [expandedKeys, setExpandedKeys] = useInternalState<Set<Key>>({
+		initialValue: new Set(),
+		onChange: props.onExpandedChange,
+		value: props.expandedKeys,
+	});
+
+	const [items, setItems] = useInternalState({
+		initialValue: [],
+		onChange: props.onItemsChange,
+		value: props.items,
+	});
+
+	const selection = useMultipleSelection({
+		onSelectionChange: props.onSelectionChange,
+		selectedKeys: props.selectedKeys,
+	});
+
+	const reorder = (from: Array<number>, path: Array<number>) => {
+		if (!Array.isArray(items)) {
+			return;
+		}
+
+		const tree = createImmutableTree(items);
+
+		tree.produce(from, path);
+
+		setItems(tree.applyPatches());
+	};
+
+	const toggle = (key: Key) => {
+		const expanded = new Set(expandedKeys);
+
+		if (expanded.has(key)) {
+			expanded.delete(key);
+		} else {
+			expanded.add(key);
+		}
+
+		setExpandedKeys(expanded);
+	};
+
+	const open = (key: Key) => {
+		const expanded = new Set(expandedKeys);
+
+		if (!expanded.has(key)) {
+			expanded.add(key);
+		}
+
+		setExpandedKeys(expanded);
+	};
+
+	return {
+		expandedKeys,
+		items,
+		open,
+		reorder,
+		selection,
+		toggle,
+	};
+}
+
+// RFC 6902 4.4
+type Patch = {
+	op: 'move';
+	from: Array<number>;
+	path: Array<number>;
+};
+
+function createImmutableTree<T extends Array<Record<string, any>>>(tree: T) {
+	const patches: Array<Patch> = [];
+
+	function nodeByPath(path: Array<number>): any {
+		const queue = [...path];
+		const rootIndex: number = queue.shift() as number;
+		let item = tree[rootIndex];
+		let parent = tree;
+		let index = rootIndex;
+
+		while (queue.length) {
+			index = queue.shift() as number;
+
+			if (item.children) {
+				parent = item.children;
+				item = item.children[index];
+			}
+		}
+
+		return {
+			index,
+			item,
+			parent,
+		};
+	}
+
+	function applyPatches(): T {
+		patches.forEach((patch) => {
+			const {from, op, path} = patch;
+
+			switch (op) {
+				// Applies the operation on the tree, the move is functionally
+				// identical to a "remove" operation on the `from` location and
+				// immediately followed by the "add" operation at the target
+				// location with the value that was removed.
+				case 'move': {
+					// The node was "moved" in the same "root node"
+					if (from[0] === path[0]) {
+						const nodeToMove = nodeByPath(from);
+						const {children} = tree[from[0]];
+						children.splice(nodeToMove.index, 1);
+						children.splice(path[1], 0, nodeToMove.item);
+					} else {
+						const nodeToRemove = nodeByPath(from);
+						const parentNode = nodeToRemove.parent.filter(
+							(item: any, index: number) =>
+								index !== nodeToRemove.index
+						);
+
+						tree[from[0]] = {
+							...tree[from[0]],
+							children: parentNode,
+						};
+
+						const {item} = nodeByPath(path);
+
+						tree[path[0]] = {
+							...item,
+							children: [...item.children, nodeToRemove.item],
+						};
+					}
```

Comment:
We need to clean up a little more here to make this work for nested elements, otherwise, it will only work on the first level of the tree.

Reply:
Yes, that said before this change I tested in the "nested" story and it was working fine (the body of the `else` block). My reasoning was the following: if you have `from` that looks like this: `[0][1]`, and you have `path` that looks like this `[0][2]`, then you know that the items are being moved in the same "root" node, but you are right; it does need some cleaning/improvement because you could also have `from` that looks like this `[0][1][0]`  and path that looks like this `[0][2][1]` - I'll look into that.

Reply:
I see, to identify when we should move into `children` or move up or down from the item, we can do a comparison of `path` and `from`, compare the size of both Arrays if they are equal, and check if the indexes are equal, so we know that it is only to reorder at the same level but we still need to identify the position of the element, it is also worth checking if the reordering in the same level is something valid to be implemented for a Tree.